pkg/webhook/admission/webhook_test.go

```diff
@@ -0,0 +1,379 @@
+				Expect(alwaysAllow.invoked).To(BeFalse())
+			})
+		})
+	})
+
+	Describe("mutating webhook", func() {
+		Context("patch handler not returning the right patch type", func() {
+			req := &http.Request{
+				Header: http.Header{"Content-Type": []string{"application/json"}},
+				Body:   nopCloser{Reader: bytes.NewBufferString(`{"request":{}}`)},
+			}
+			errPatcher := &fakeHandler{
+				fn: func(ctx context.Context, req Request) Response {
+					return Response{
+						Patches: []jsonpatch.JsonPatchOperation{
+							{
+								Operation: "add",
+								Path:      "/metadata/annotation/new-key",
+								Value:     "new-value",
+							},
+						},
+						Response: &admissionv1beta1.AdmissionResponse{
+							Allowed: true,
+						},
+					}
+				},
+			}
+			wh := &Webhook{
+				Type:     types.WebhookTypeMutating,
+				Handlers: []Handler{errPatcher},
+				KVMap:    map[string]interface{}{"foo": "bar"},
+			}
+			w := &fakeResponseWriter{}
+			expected := []byte(
+				`{"response":{"uid":"","allowed":false,"status":{"metadata":{},"message":"unexpected patch type ` +
+					`returned by the handler: \u003cnil\u003e, only allow: JSONPatch","code":500}}}
+`)
+			It("should return a response with error", func() {
+				wh.ServeHTTP(w, req)
+				Expect(w.response).NotTo(BeNil())
+				Expect(w.response).To(Equal(expected))
+				Expect(errPatcher.invoked).To(BeTrue())
+				Expect(errPatcher.valueFromContext).To(Equal("bar"))
+			})
+		})
+
+		Context("multiple patch handlers", func() {
+			req := &http.Request{
+				Header: http.Header{"Content-Type": []string{"application/json"}},
+				Body:   nopCloser{Reader: bytes.NewBufferString(`{"request":{}}`)},
+			}
+			patcher1 := &fakeHandler{
+				fn: func(ctx context.Context, req Request) Response {
+					return Response{
+						Patches: []jsonpatch.JsonPatchOperation{
+							{
+								Operation: "add",
+								Path:      "/metadata/annotation/new-key",
+								Value:     "new-value",
+							},
+							{
+								Operation: "replace",
+								Path:      "/spec/replicas",
+								Value:     "2",
+							},
+						},
+						Response: &admissionv1beta1.AdmissionResponse{
+							Allowed:   true,
+							PatchType: func() *admissionv1beta1.PatchType { pt := admissionv1beta1.PatchTypeJSONPatch; return &pt }(),
+						},
+					}
+				},
+			}
+			patcher2 := &fakeHandler{
+				fn: func(ctx context.Context, req Request) Response {
+					return Response{
+						Patches: []jsonpatch.JsonPatchOperation{
+							{
+								Operation: "add",
+								Path:      "/metadata/annotation/hello",
+								Value:     "world",
+							},
+						},
+						Response: &admissionv1beta1.AdmissionResponse{
+							Allowed:   true,
+							PatchType: func() *admissionv1beta1.PatchType { pt := admissionv1beta1.PatchTypeJSONPatch; return &pt }(),
+						},
+					}
+				},
+			}
+			wh := &Webhook{
+				Type:     types.WebhookTypeMutating,
+				Handlers: []Handler{patcher1, patcher2},
```

Comment:
Do you check conflict among different handlers?

Reply:
No. I can't find a easy way to do that.
I added a comment in the struct saying that user should ensure that their handlers should not mutate same field and generate conflicting JSON patches.